### PR TITLE
Projects

### PR DESCRIPTION
**Pull request by tiferrei**
Here is the PR containing the repo –> project changes, at the moment is failing because I wasn't able to merge or rebase from master, to include the commit requests, required by the tests.

~~PS: I have no idea how to merge this. `project.pbxproj` is pure hell when it comes to merge conflicts.~~

Nobody's born knowing 👍

### CODE REVIEW

**Comment by pietbrauer** [comment]
Tests seem to fail for macOS:

```bash
CommitTests
❌  fatal error: unexpectedly found nil while unwrapping an Optional value
```

**Comment by tiferrei** [comment]
Fixed, after this one I'll just open the one for the `login` –> `username` changes, and after that it's pretty much ready for major release?

File: TanukiKit/Project.swift [hunk 1]
```diff
@@ -98,6 +98,71 @@ public enum VisibilityLevel: Int {
...
+            self.id = id
+            if let urlString = json["url"] as? String, let parsedURL = URL(string: urlString) {
+                url = parsedURL
+            }
+            projectID = json["project_id"] as? Int
+            pushEvents = json["push_events"] as? Bool
+            issuesEvents = json["issues_events"] as? Bool
+            mergeRequestsEvents = json["merge_requests_events"] as? Bool
+            tagPushEvents = json["tag_push_events"] as? Bool
+            noteEvents = json["ntoe_events"] as? Bool
```
Comment (pietbrauer) [inline TanukiKit/Project.swift]: Small typo here `ntoe_events` should be `note_events` (`projectHooks` is missing tests)

File: TanukiKit/Project.swift [hunk 2]
```diff
@@ -174,9 +239,75 @@ public enum VisibilityLevel: Int {
...
+        userName = json["user_name"] as? String
+        userEmail = json["user_email"] as? String
+        if let urlString = json["user_avater"] as? String, let urlFromString = URL(string: urlString) {
+            userAvatar = urlFromString
+        }
+        projectID = json["project_id"] as? Int
+        project = Project(json["project"] as? [String: AnyObject] ?? [:])
+        commits =  (json["commits"] as? [[String: AnyObject]])?.map { Commit($0) }
+        totalCommitsCount = json["total_commits_count"] as? Int
+        //repository = Repository(json["repository"] as? [String: AnyObject] ?? [:])
```
Comment (pietbrauer) [inline TanukiKit/Project.swift]: Please remove the dead code

File: TanukiKitTests/ProjectTests.swift [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        XCTAssertEqual(event.targetType, nil)
+        XCTAssertEqual(event.authorID, 8)
+        XCTAssertEqual(event.data!.objectKind, "push")
+        XCTAssertEqual(event.data!.eventName, "push")
+        XCTAssertEqual(event.data!.before, "7a6a450ca607c9c0be539bb91c5c7eb6bdf4cc79")
+        XCTAssertEqual(event.data!.after, "bde8115a0d77c6a0bcad8e427ecbc38c4a6a4f5f")
+        XCTAssertEqual(event.data!.ref, "refs/heads/master")
+        print("DEBUG: \(event.data!.commits)")
+        XCTAssertEqual(event.data!.commits![0].message, "The private token is no longer given\n")
+        // TODO: Finish Parsing tests
```
Comment (pietbrauer) [inline TanukiKitTests/ProjectTests.swift]: Is this still in Progress?

File: TanukiKitTests/ProjectTests.swift [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+        XCTAssertEqual(event.actionName, "pushed to")
+        XCTAssertEqual(event.targetID, nil)
+        XCTAssertEqual(event.targetType, nil)
+        XCTAssertEqual(event.authorID, 8)
+        XCTAssertEqual(event.data!.objectKind, "push")
+        XCTAssertEqual(event.data!.eventName, "push")
+        XCTAssertEqual(event.data!.before, "7a6a450ca607c9c0be539bb91c5c7eb6bdf4cc79")
+        XCTAssertEqual(event.data!.after, "bde8115a0d77c6a0bcad8e427ecbc38c4a6a4f5f")
+        XCTAssertEqual(event.data!.ref, "refs/heads/master")
+        print("DEBUG: \(event.data!.commits)")
```
Comment (pietbrauer) [inline TanukiKitTests/ProjectTests.swift]: Please remove the debug `print`

**Review by pietbrauer — CHANGES_REQUESTED** [review]
Thanks for the PR! Just some small things. Please make sure tests war written for the new endpoints.

File: TanukiKitTests/ProjectTests.swift (hunk 3 above)
Comment (tiferrei) [reply TanukiKitTests/ProjectTests.swift]: Oh I wasn't completely sure wether I should do those parsing tests as they just repeat the commit and event parsing tests? What would you prefer me to do?

Comment (pietbrauer) [reply TanukiKitTests/ProjectTests.swift]: Given that [there was an error](https://github.com/nerdishbynature/TanukiKit/pull/33#discussion_r102021287) in the repetition I would prefer to have tests for that.

**Comment by tiferrei** [comment]
```log
2017-02-22 22:40:57.828 xcodebuild[16482:31076] Error Domain=IDETestOperationsObserverErrorDomain Code=3 "Test operation was canceled. If you believe this error represents a bug, please attach the log file at /Users/travis/Library/Developer/Xcode/DerivedData/TanukiKit-fvezsfumqfjrojegffuqyyeakarq/Logs/Test/B5D72806-DA2D-4705-AD6E-CB903AB72D8B/Session-TanukiKitTests-2017-02-22_224056-ZN9N3c.log" UserInfo={NSLocalizedDescription=Test operation was canceled. If you believe this error represents a bug, please attach the log file at /Users/travis/Library/Developer/Xcode/DerivedData/TanukiKit-fvezsfumqfjrojegffuqyyeakarq/Logs/Test/B5D72806-DA2D-4705-AD6E-CB903AB72D8B/Session-TanukiKitTests-2017-02-22_224056-ZN9N3c.log}
```

@pietbrauer Travis seems to be having a hard time with this one, took it hours to start and then failed within 5 minutes?

Update: I just ran the CI job locally, with `bundle`, and everything works perfectly, don't know what's wrong with Travis.

**Comment by tiferrei** [comment]
@pietbrauer: Am I seeing things, or did that space in the end of the declaration mess up the whole CI process?

**Comment by pietbrauer** [comment]
No, it messed up Xcode. I just opened the project in Xcode, clicked on the "Fix it" button and the build error went away.

**Comment by tiferrei** [comment]
Hmm interesting, everything ran perfectly on mine? Even the CI commands.

**Review by pietbrauer — APPROVED** [review]
Great work, don't know why the Xcode Project has so many changes but I'm ok with it.

**Pull request merged**